opt, cat: cleaner API for mutation columns

RaduBerinde opened the pull request:
**NOTE** I used `AllColumnCount` so that the compiler forces me to look at all callsites, and left it like this so they are also apparent in the diff. I intend to rename it to `ColumnCount` before merging.

The catalog currently relies on a particular ordering to separate
different kinds of columns (ordinary, write-only, delete-only). This
is unfortunate because it restricts the underlying implementation; in
particular, it makes it tricky to add a new kind of column (like the
`mvcc_timestamp` system column) without potentially breaking existing
code in subtle ways.

This change reworks the cat API to have a single `ColumnCount` method
and a separate `ColumnKind` method used to figure out the column kind.
This forces all code that cares about the kind to check it explicitly
rather than relying on a potentially faulty column count.

Fixes #51323.

Release note: None

RaduBerinde committed:
opt, cat: cleaner API for mutation columns

The catalog currently relies on a particular ordering to separate
different kinds of columns (ordinary, write-only, delete-only). This
is unfortunate because it restricts the underlying implementation; in
particular, it makes it tricky to add a new kind of column (like the
`mvcc_timestamp` system column) without potentially breaking existing
code in subtle ways.

This change reworks the cat API to have a single `ColumnCount` method
and a separate `ColumnKind` method used to figure out the column kind.
This forces all code that cares about the kind to check it explicitly
rather than relying on a potentially faulty column count.

Fixes #51323.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/51400)
<!-- Reviewable:end -->

andy-kimball commented on pkg/sql/opt/cat/utils.go:
```diff
@@ -122,15 +122,15 @@ func ResolveTableIndex(
 // ConvertColumnIDsToOrdinals converts a list of ColumnIDs (such as from a
-// tree.TableRef), to a list of ordinal positions of columns within the given
-// table. See tree.Table for more information on column ordinals.
+// tree.TableRef), to a list of ordinal positions of non-mutation columns within
```
It seems like an API bug that this only looks at non-mutation cols. Why not let it handle mutation cols as well, since the API is generic? I don't think any existing caller calls it with mutation cols, so there should be no panics.

andy-kimball commented on pkg/sql/opt/cat/utils.go:
```diff
@@ -147,8 +147,8 @@ func ConvertColumnIDsToOrdinals(tab Table, columns []tree.ColumnID) (ordinals []
 // FindTableColumnByName returns the ordinal of the non-mutation column having
 // the given name, if one exists in the given table. Otherwise, it returns -1.
 func FindTableColumnByName(tab Table, name tree.Name) int {
```
Should we make this return mutation cols as well and require caller to test for mutation/system cols? The more col types we add, the more confusing it is when some generic-sounding APIs filter cols and others don't.

andy-kimball commented on pkg/sql/opt/memo/logical_props_builder.go:
```diff
@@ -1268,8 +1268,8 @@ func (b *logicalPropsBuilder) buildMutationProps(mutation RelExpr, rel *props.Re
-	for i, n := 0, tab.ColumnCount(); i < n; i++ {
-		if private.IsColumnOutput(i) {
+	for i, n := 0, tab.AllColumnCount(); i < n; i++ {
+		if !cat.IsMutationColumn(tab, i) && private.IsColumnOutput(i) {
```
Actually, I think we can just get rid of the comment and the check for `IsMutationColumn`. `private.IsColumnOutput` should always return false for mutation columns, or else there's a bug somewhere else in the code. We always set `private.ReturnCols` entries to zero for mutation cols.

andy-kimball commented on pkg/sql/opt/memo/logical_props_builder.go:
```diff
@@ -1289,7 +1289,10 @@ func (b *logicalPropsBuilder) buildMutationProps(mutation RelExpr, rel *props.Re
 	// the column must be not null.
-	for i, n := 0, tab.ColumnCount(); i < n; i++ {
+	for i, n := 0, tab.AllColumnCount(); i < n; i++ {
+		if cat.IsMutationColumn(tab, i) {
```
The `OutputCols` check should already take care of this case, since mutation cols will never be output cols by the check above. I think it's good to let the mutation builder decide which cols are output or not, and let this props builder code just blindly process the cols without checking again.

andy-kimball commented on pkg/sql/opt/memo/multiplicity_builder.go:
```diff
@@ -118,8 +118,10 @@ func deriveUnfilteredCols(in RelExpr) opt.ColSet {
-			for i, cnt := 0, baseTable.ColumnCount(); i < cnt; i++ {
-				unfilteredCols.Add(t.Table.ColumnID(i))
+			for i, cnt := 0, baseTable.AllColumnCount(); i < cnt; i++ {
+				if !cat.IsMutationColumn(baseTable, i) {
```
I don't see any harm in adding mutation cols here. I think Drew just used `ColumnCount` b/c it seemed like the right thing. But really we want all columns.

andy-kimball commented on pkg/sql/opt/memo/statistics_builder_test.go:
```diff
@@ -99,7 +99,7 @@ func TestGetStatsFromConstraint(t *testing.T) {
 
 		var cols opt.ColSet
-		for i := 0; i < tab.ColumnCount(); i++ {
+		for i := 0; i < tab.AllColumnCount(); i++ {
```
I wonder how mutation cols play with stats. A good example of a case where using `ColumnCount` can cause us to not consider a case that is potentially interesting to think about, or at least document.

andy-kimball commented on pkg/sql/opt/optbuilder/insert.go:
```diff
@@ -1054,9 +1057,9 @@ func (mb *mutationBuilder) mapColumnNamesToOrdinals(names tree.NameList) util.Fa
 	var ords util.FastIntSet
```
Can you add to the `mapColumnNamesToOrdinals` to make it clear that it skips mutation columns?

andy-kimball commented on pkg/sql/opt/optbuilder/mutation_builder.go:
```diff
@@ -825,11 +829,14 @@ func (mb *mutationBuilder) makeMutationPrivate(needResults bool) *memo.MutationP
 
 	if needResults {
-		// Only non-mutation columns are output columns. ReturnCols needs to have
 		// DeletableColumnCount entries, but only the first ColumnCount entries
```
Need to update the comment to remove ref to `DeletableColumnCount`.

andy-kimball commented on pkg/sql/opt/optbuilder/select.go:
```diff
@@ -577,8 +578,8 @@ func (b *Builder) addCheckConstraintsForTable(tabMeta *opt.TableMeta) {
 
 	// Find the non-nullable table columns.
```
Add parenthetical to the comment, saying that mutation cols can be NULL during backfill, so they should be excluded.

andy-kimball commented on pkg/sql/opt/optbuilder/select.go:
```diff
@@ -626,9 +627,9 @@ func (b *Builder) addCheckConstraintsForTable(tabMeta *opt.TableMeta) {
 	var tableScope *scope
 	tab := tabMeta.Table
-	for i, n := 0, tab.ColumnCount(); i < n; i++ {
+	for i, n := 0, tab.AllColumnCount(); i < n; i++ {
```
Add comment to explain why mutation cols should not be included here.

andy-kimball reviewed (approved):
LGTM, with some nits.

rohany commented:
From an outside perspective, I like this a lot. It's much more explicit about what count to use/what columns are considered in an operation, and will also make it easier for me to catch all of the cases where we should or shouldn't include system columns.

RaduBerinde committed:
opt: move ConvertColumnIDsToOrdinals to the optbuilder

Also add some tests that fail if we don't ignore mutation columns.

Release note: None

RaduBerinde committed:
opt: move FindTableColumnBYName to optbuilder

RaduBerinde reviewed (commented):
Updated.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51400#-:-MCCz-_a4gvBuJiMvN5q:b-bw8e14)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @mgartner, @RaduBerinde, @rohany, and @rytaft)

___
*[pkg/sql/opt/cat/utils.go, line 125 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51400#-MC9gHghwYWvwRVMV6IZ-r1-125:-MCCkxAk0FOR9a9gee-f:b8ii18c) ([raw file](https://github.com/cockroachdb/cockroach/blob/3023579dea0269a31f5715d7994ec76fcdab2621/pkg/sql/opt/cat/utils.go#L125)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

It seems like an API bug that this only looks at non-mutation cols. Why not let it handle mutation cols as well, since the API is generic? I don't think any existing caller calls it with mutation cols, so there should be no panics.
</blockquote></details>

The caller does care - this is used when referencing a table by ID. Without the check, we allow a mutation column (which in practice leads to a more obscure error a bit later). I added some tests and I moved this into the optbuilder as it's specific to building numeric references.
___
*[pkg/sql/opt/cat/utils.go, line 149 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51400#-MC9gHghwYWvwRVMV6IZ-r1-149:-MCCv1134FKYijxHW4ek:bqua9qm) ([raw file](https://github.com/cockroachdb/cockroach/blob/3023579dea0269a31f5715d7994ec76fcdab2621/pkg/sql/opt/cat/utils.go#L149)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Should we make this return mutation cols as well and require caller to test for mutation/system cols? The more col types we add, the more confusing it is when some generic-sounding APIs filter cols and others don't.
</blockquote></details>

Moved to optbuilder, renamed to `findPublicTableColumnByName`.
___
*[pkg/sql/opt/memo/logical_props_builder.go, line 1272 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51400#-MC9gHgi8O6FySWojH6t-r1-1272:-MCCvpAu69bzwZEbJto9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3023579dea0269a31f5715d7994ec76fcdab2621/pkg/sql/opt/memo/logical_props_builder.go#L1272)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Actually, I think we can just get rid of the comment and the check for `IsMutationColumn`. `private.IsColumnOutput` should always return false for mutation columns, or else there's a bug somewhere else in the code. We always set `private.ReturnCols` entries to zero for mutation cols.
</blockquote></details>

Done.
___
*[pkg/sql/opt/memo/logical_props_builder.go, line 1293 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51400#-MC9gHgi8O6FySWojH6t-r1-1293:-MCCvylL071bGaJVklrY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3023579dea0269a31f5715d7994ec76fcdab2621/pkg/sql/opt/memo/logical_props_builder.go#L1293)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

The `OutputCols` check should already take care of this case, since mutation cols will never be output cols by the check above. I think it's good to let the mutation builder decide which cols are output or not, and let this props builder code just blindly process the cols without checking again.
</blockquote></details>

Done.
___
*[pkg/sql/opt/memo/multiplicity_builder.go, line 122 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51400#-MC9gHgi8O6FySWojH6u-r1-122:-MCCw877471b-TmCI5fj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3023579dea0269a31f5715d7994ec76fcdab2621/pkg/sql/opt/memo/multiplicity_builder.go#L122)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I don't see any harm in adding mutation cols here. I think Drew just used `ColumnCount` b/c it seemed like the right thing. But really we want all columns.
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/insert.go, line 1057 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51400#-MC9gHgi8O6FySWojH6y-r1-1057:-MCCwlycALt4mLuuOu8X:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3023579dea0269a31f5715d7994ec76fcdab2621/pkg/sql/opt/optbuilder/insert.go#L1057)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Can you add to the `mapColumnNamesToOrdinals` to make it clear that it skips mutation columns?
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 832 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51400#-MC9gHgi8O6FySWojH6z-r1-832:-MCCwtcJ06Vwgi11W6ig:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3023579dea0269a31f5715d7994ec76fcdab2621/pkg/sql/opt/optbuilder/mutation_builder.go#L832)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Need to update the comment to remove ref to `DeletableColumnCount`.
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/select.go, line 579 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51400#-MC9gHgj2qN97IDexWuM-r1-579:-MCCxrVh0IS5wSTw7AN1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3023579dea0269a31f5715d7994ec76fcdab2621/pkg/sql/opt/optbuilder/select.go#L579)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Add parenthetical to the comment, saying that mutation cols can be NULL during backfill, so they should be excluded.
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/select.go, line 630 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51400#-MC9gHgj2qN97IDexWuM-r1-630:-MCCyIey49pJUuDxKhgN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3023579dea0269a31f5715d7994ec76fcdab2621/pkg/sql/opt/optbuilder/select.go#L630)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Add comment to explain why mutation cols should not be included here.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
This is great! I think we currently do something similar with the Index catalog, so it'd be great to apply similar changes there at some point.

> Reviewed 10 of 20 files at r1, 4 of 8 files at r2, 3 of 4 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51400#-:-MCDOkmD9EVtTeDvle3U:b6jyp14)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @mgartner, @RaduBerinde, @rohany, and @rytaft)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+

RaduBerinde commented:
bors r+
